### ISSUE TITLE
`extend` and `to_interval_set` panic if input isn't sorted

### ISSUE THREAD

**Issue by lawbel** [ptal/intervallum#25]
Minimal working example:

```rust
use gcollections::ops::*;
use interval::interval_set::*;
use interval::interval::*;

fn main() {
    let pairs = vec![
        (2, 6),
        (1, 5),
    ];

    // panic #1 happens here
    let panic1 = pairs.clone().to_interval_set();

    let mut panic2 = IntervalSet::empty();
    // panic #2 happens here
    panic2.extend(pairs.iter().map(|pair| pair.to_interval()));

    println!("{}", panic1);
    println!("{}", panic2);
}
```

If you run this example, you'll see that you get a panic from the call to `to_interval_set`. If you comment out that line, you'll find a second panic caused by the call to `extend`.

I implied in the issue title that the input needs to be sorted, but actually that's not _exactly_ right. The first element of the pairs needs to be sorted, but the second elements can come in any order - so `vec![(1, 6), (1, 4)]` will not cause a panic, but `vec![(6, 1), (4, 1)]` would.

I've had a look into the source, and it seems that `to_interval_set` iterates through it's argument, each time calling `extend` to insert the elements. So the issue is that that calling `to_interval_set` on a `Vec<(Bound, Bound)>` (or calling `extend` on an iterator) ends up calling the internal function `join_or_push` which expects the first elements to come in order.

I'd be happy to attempt to put together a PR to fix this if that would be welcome?

**Comment by ptal** [comment]
Interesting! Thank you for reporting this bug. I'd welcome a PR.
For info (although I'm sure you already know that): the list `[(2, 6), (1, 5)]` should produce the interval set `[(1,6)]` and `vec![(6, 1), (4, 1)]` should give an empty interval set.

Also, we might still need an (probably private) operation that does not sort the elements, I'm not sure about that, but set operations such as union and intersection might use extend "the right way" with elements already sorted; so sorting a second time would be useless.

**Comment by lawbel** [comment]
@ptal ah yes, I got so caught up in digging into the code that I almost forgot we're working with intervals and gave a strange example of `vec![(6, 1), (4, 1)]`. I'll work on a PR, try to get something reasonable put together in the next few days.

From skimming the codebase it seemed like `extend` was only used in two places - exposed as part of the public interface, and then called in the implementation of `to_interval_set`. In any case, perhaps it would be best to have an internal helper function (call it something like `extend_unsafe`) which does what the current `extend` function does (so it does no sorting, it relies on the invariant being upheld of the elements being in the right order). Then the user-facing functions `extend` and `to_interval_set` could do a sort and then call `extend_unsafe` to do the rest.

**Comment by ptal** [comment]
Sounds nice! Probably no need for `extend_unsafe` if it is not used anywhere else. Thanks.

**Comment by ptal** [comment]
Fixed by #26
